Incorporate input from IETF 114 (#43)
intarchboard/draft-protocol-maintenance: draft-iab-protocol-maintenance.md

without aborting unrelated processing.

Robustness to attacks:

: Since not all actors on the Internet are benevolent, networking software needs
to be resilient to input that is intentionally crafted to cause unexpected
consequences. For example, software must ensure that invalid input doesn't allow
the sender to access data that it would otherwise not be allowed to.

Robustness to the unexpected:

: It can be possible for an implementation to receive inputs that the
specification did not prepare it for. This interpretation covers the scenario
where a specific input is explicitly defined to be faulty (or if there is
ambiguity in the specification regarding whether the input is valid), but where
the specification does not explicitly specify how to react to that input. In
this case, the robustness principle advocates that the implementation tolerate
the faulty input and silently discard it.

While the first two interpretations above are understood to be correct and
beneficial, the third interpretation is no longer considered best practice in
all scenarios. The rest of this document uses the term "robustness principle" to

> refer to this third interpretation.

Author: Sounds good, I tried to do that but found too many instances. I agree that giving this interpretation a different name would be the best solution, assuming we come up with a good name. (I haven't found one yet)

  deployments are tolerant of the same non-compliant behavior.

This happens because interoperability requirements for protocol implementations
are set by other deployments. The IETF does not create nor mandate conformance
test suites for protocols, causing established implementations to become

> de-facto conformance test suites used by new implementations.

Author: This was a comment from Stuart that I had written a while back in #41. I've removed it from this PR and we can discuss it there since it's orthogonal.

For a protocol to have sustained viability, it is necessary for both
specifications and implementations to be responsive to changes, in addition to
handling new and old problems that might arise over time. When an implementor
discovers a scenario where a specification defines some input as faulty but does
not define how to handle that input, the implementor can provide significant
value to the ecosystem by reporting the issue and helping evolve the

> specification.

Author: Yeah adding For example makes sense - done

HTTP/2 {{H2}} resulted in improvements in the security of the
deployed base.

> When the protocol involves shared state between multiple parties (often refered
> to as a connection), the fatal errors described in this section are best sent
> over the network as explicit error messages. Additionally, those error messages

Reviewer: I don't think that this depends on shared protocol state.  ICMP defines exactly the sort of feedback you describe here, where the offending party is informed of what they did that was "wrong".

```suggestion
A notification for a fatal error is best as explicit error messages to the entity that
made the error. Error messages
```

> benefit from being extensible and carrying debugging information to help the

Reviewer: ```suggestion
benefit from being able to carry arbitrary information that might help the
```

implementor of the sender of the faulty input understand and fix the issue in
their software. QUIC error frames {{?QUIC=RFC9000}} are an example of a fatal
error mechanism that helped implementors improve software quality throughout the
protocol lifecycle.

> Conversely, when the faulty input involves a single packet received from a
> non-authenticated sender, the associated fatal error need not be sent over the
> wire: the receiver can simply abort processing and silently discards the packet
> to avoid denial-of-service and amplification attacks.

Reviewer: ```suggestion
Stateless protocol endpoints might generate denial-of-service attacks if they
respond with an error messages in response to every message that is received
from an unauthenticated sender.  These implementations might need to silently
discard these messages.
```
